### PR TITLE
Improve memory/hashing performance

### PR DESCRIPTION
**Pull request by Yard1**
Signed-off-by: Antoni Baum <antoni.baum@protonmail.com>

# Related Issue or bug

Info about Issue or bug

Closes #[issue number that will be closed through this PR]

# Describe the changes you've made

Changes to the joblib hashing/memory include:
1. Using the xxhash algorithm by default instead of md5.
2. Using the highest available pickle protocol by default.
3. Using O(1) hashing (https://github.com/joblib/joblib/pull/1011).
4. Special support for pandas/numpy object arrays.
5. Caching the function output only if the call takes more than 0.1 seconds (this increases runtime slightly, but drastically reduces the disk space used).
6. Avoiding hashing the function signature twice in `Memory` class.

In addition, this PR:
* Patches several pyod models to be cloneable (necessary for hashing/caching), upstream tracked https://github.com/yzhao062/pyod/pull/471
* Caches the output of all pipeline transformation steps, providing a fastpath for long pipelines
* Ensures final estimators in pipelines are cached correctly

Benchmark consists of running setup on a dataset with `remove_multicollinearity`, `feature_selection`, `pca` and `transform`, then creating 4 dummy models and accessing `dataset_transformed`, `train_transformed` and `test_transformed` 8 times each. Results (min of 3 runs) on Python 3.8, AWS t3.2xlarge instance - the number in the brackets is the difference between original algorithm and new algorithm (smaller = better for new algorithm) with "Caching the function output only if the call takes more than 0.1 seconds" turned off. Benchmark code can be found in `tests/benchmarks/test_memory_performance.py::_test_e2e`
```
(diamond) Original: 44.52244117693044 vs PyCaret: 24.960464466013946 (-19.561976710916497)
(wine) Original: 17.64841262390837 vs PyCaret: 13.987546702031977 (-3.6608659218763933)
(CTG) Original: 45.246672917040996 vs PyCaret: 29.04680022096727 (-16.199872696073726)
(house) Original: 116.16553873103112 vs PyCaret: 76.77832635794766 (-39.38721237308346)
(parkinsons) Original: 17.494275477947667 vs PyCaret: 14.636485776980408 (-2.8577897009672597)
(electrical_grid) Original: 16.84290632500779 vs PyCaret: 14.334434765973128 (-2.5084715590346605)
(employee) Original: 32.556646017939784 vs PyCaret: 21.32283796800766 (-11.233808049932122)
(telescope) Original: 18.44674307096284 vs PyCaret: 15.701882674940862 (-2.744860396021977)
(bike) Original: 29.071959833963774 vs PyCaret: 19.954391069011763 (-9.117568764952011)
(gold) Original: 44.764836048008874 vs PyCaret: 30.723237924976274 (-14.0415981230326)
(traffic) Original: 72.42423428199254 vs PyCaret: 37.58492319402285 (-34.83931108796969)
(income) Original: 193.36588983400725 vs PyCaret: 61.884934403933585 (-131.48095543007366)
(credit) Original: 24.960413201944903 vs PyCaret: 19.28032624803018 (-5.680086953914724)
(bank) Original: 221.88767109590117 vs PyCaret: 61.53595164499711 (-160.35171945090406)
(poker) Original: 63.8842476970749 vs PyCaret: 51.40233268798329 (-12.481915009091608)
```

# Type of change

Please delete options that are not relevant.

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Code style update (formatting, local variables)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update

# How Has This Been Tested?

Please describe the tests that you ran to verify your changes. Provide instructions so we can reproduce. 

# Describe if there is any unusual behaviour of your code(Write `NA` if there isn't)

A clear and concise description of it.

# Checklist:

- [ ] My code follows the style guidelines of this project.
- [ ] I have performed a self-review of my own code.
- [ ] I have commented my code, particularly in hard-to-understand areas.
- [ ] I have made corresponding changes to the documentation.
- [ ] My changes generate no new warnings.
- [ ] I have added tests that prove my fix is effective or that my feature works.
- [ ] New and existing unit tests pass locally with my changes.
- [ ] Any dependent changes have been merged and published in downstream modules.

# Screenshots

 Original           | Updated
 :--------------------: |:--------------------:
 **original screenshot**  | <b>updated screenshot </b> |

### CODE REVIEW

File: tests/test_memory.py [hunk 1]
```diff
@@ -0,0 +1,336 @@
...
+    time_func1 = time_func(func1, *args)
+    time_func2 = time_func(func2, *args)
+    if time_func1 + time_func2 == 0:
+        return 0
+    relative_diff = 0.5 * (abs(time_func1 - time_func2) / (time_func1 + time_func2))
+    return relative_diff
+
+
+@pytest.mark.parametrize("dtype", ("int", "category", "object"))
+def test_pandas_dataframe(dtype):
```
Comment (tvdboom) [inline tests/test_memory.py:53]: why are we testing this? Isn't this outside of pycaret's scope?

**Review by tvdboom — COMMENTED** [review]
looks good. just wondering if all that testing is necessary. It seems to me that you are testing stuff unrelated to this library.

File: tests/test_memory.py (hunk 1 above)
Comment (Yard1) [reply tests/test_memory.py:53]: It is necessary as we have added new hashing logic that is not tested in joblib! The test suite is minimal, and benchmarks can be turned off if needed.

**Review by tvdboom — APPROVED** [review]

File: tests/test_memory.py (hunk 1 above)
Comment (tvdboom) [reply tests/test_memory.py:53]: alright. accepted

**Pull request merged**